### PR TITLE
`cargo install` will ignore the target triple specified in a project directory

### LINKED ISSUE #5441: `cargo install` uses my project's `.cargo/config` target triple and it shouldn't?
**TL;DR:** I spent multiple hours trying to figure out why I couldn't `cargo install` some tools. It turned out that the reason I couldn't was because, unbeknownst to me, `cargo install` was trying to build the tools for the wrong target due to `.cargo/config` in my project directory!

### Full explanation

1. My project has a `.cargo/config` file to build for a wasm target triple since my project is wasm only:

```
[build]
target = "wasm32-unknown-emscripten"
```

2. I ran `cargo install cargo-web` from my project's working directory. It failed due to a bunch of obscure errors about dependencies being unable to compile. **This made me think the dependencies had been really broken, which gave me a really sour taste in my mouth for Rust.**

3. After cloning the dependencies [and trying to fix them](https://github.com/softprops/atty/issues/26), I finally realized: `cargo install cargo-web` is for _some reason_ building for `wasm32-unknown-emscripten` on my system! 😢

As it turned out, `cargo install` is respecting my project's `.cargo/config` target triple.

### What I expected

I think that `cargo install` should _ignore_ the target triple specified in a project directory.

I imagine that at some point or another, most users developing projects primarily targeting wasm/asmjs will run into this issue, and it is _incredibly_ annoying to debug.

I'm using:

```
$ cargo --version
cargo 1.27.0-nightly (0a1add2d8 2018-04-20)
```

### PR DESCRIPTION
**Pull request by vramana**
Fixes #5441

### CODE REVIEW

File: src/bin/cargo/commands/install.rs [hunk 1]
```diff
@@ -72,8 +72,13 @@ continuous integration systems.",
 }
 
 pub fn exec(config: &mut Config, args: &ArgMatches) -> CliResult {
+    let ws = args.workspace(config)?;
+    let host = config.rustc(Some(&ws)).map(|rustc| rustc.host).ok();
     let mut compile_opts = args.compile_options(config, CompileMode::Build)?;
     compile_opts.build_config.release = !args.is_present("debug");
+    // We override target architecture to host architecture since it may be
+    // set to some other architecture in .cargo/config.
+    compile_opts.build_config.requested_target = host;
```
Comment (matklad) [inline src/bin/cargo/commands/install.rs]: This could be just `.requesetd_target = None` I think

**Comment by matklad** [comment]
The test should go into https://github.com/rust-lang/cargo/blob/d33c65cbd9d6f7ba1e18b2cdb85fea5a09973d3b/tests/testsuite/install.rs directory. 

I think the test should write a `.cargo/config` file inside the project root with 

```
[build]
target = "non-existing-target"
```

and check that `cargo install` works.

File: src/bin/cargo/commands/install.rs (hunk 1 above)
Comment (vramana) [reply src/bin/cargo/commands/install.rs]: Let me change that and see if that works.

**Comment by vramana** [comment]
@matklad Sorry for the delay. I have added the test. Can you check now?

**Comment by matklad** [comment]
@alexcrichton it's interesting that we have https://github.com/rust-lang/cargo/pull/5614 as well, which is sort-of the opposite of this pull request.

**Comment by alexcrichton** [comment]
Er oops, should have taken more time to catch up! I'll continue discussion on https://github.com/rust-lang/cargo/pull/5606

**Comment by alexcrichton** [comment]
er, on https://github.com/rust-lang/cargo/pull/5614

**Pull request merged**